rotkehlchen/chain/evm/decoding/pendle/decoder.py

```diff
@@ -1,122 +1,317 @@
     DEFAULT_DECODING_OUTPUT,
+    ActionItem,
     DecoderContext,
     DecodingOutput,
 )
 from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
 from rotkehlchen.constants.assets import A_ETH
 from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
 from rotkehlchen.logging import RotkehlchenLogsAdapter
-from rotkehlchen.serialization.deserialize import deserialize_timestamp
 from rotkehlchen.types import ChecksumEvmAddress
-from rotkehlchen.utils.mixins.customizable_date import CustomizableDateMixin
-
-from .constants import CPT_PENDLE, NEW_LOCK_POSITION_TOPIC, WITHDRAW_TOPIC
-
-if TYPE_CHECKING:
-    from rotkehlchen.assets.asset import Asset
-    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
-    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
-    from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.misc import bytes_to_address
 
+from .constants import (
+    ADD_LIQUIDITY_TOPIC,
+    CPT_PENDLE,
+    KYBERSWAP_SWAPPED_TOPIC,
+    ODOS_SWAP_TOPIC,
+    OKX_ORDER_RECORD_TOPIC,
+    PENDLE_ROUTER_ABI,
+    PENDLE_ROUTER_ADDRESS,
+    PENDLE_SWAP_ADDRESS,
+    REMOVE_LIQUIDITY_TOPIC,
+    SWAP_SINGLE_TOPIC,
+    SWAP_TOKEN_FOR_PT_TOPIC,
+    SWAP_TOKEN_FOR_YT_TOPIC,
+)
 
 logger = logging.getLogger(__name__)
 log = RotkehlchenLogsAdapter(logger)
 
 
-class PendleCommonDecoder(DecoderInterface, CustomizableDateMixin):
-
-    def __init__(
-            self,
-            evm_inquirer: 'EvmNodeInquirer',
-            base_tools: 'BaseDecoderTools',
-            msg_aggregator: 'MessagesAggregator',
-            pendle_token: 'Asset',
-            ve_pendle_contract: ChecksumEvmAddress,
-    ) -> None:
-        super().__init__(
-            evm_inquirer=evm_inquirer,
-            base_tools=base_tools,
-            msg_aggregator=msg_aggregator,
+class PendleCommonDecoder(DecoderInterface):
+
+    def _decode_pendle_events(self, context: DecoderContext) -> DecodingOutput:
+        """This method decodes the following pendle events:
+        1. Buy or sell principal token(PT) with/for yield token (e.g. sUSDe, LBTC)
+        2. Buy or sell yield token(YT) with/for yield token (e.g. sUSDe, LBTC)
+        3. Adding or removing liquidity.
+        """
+        if context.tx_log.topics[0] in (SWAP_TOKEN_FOR_PT_TOPIC, SWAP_TOKEN_FOR_YT_TOPIC):
+            return self._decode_pt_yt_events(context)
+        elif context.tx_log.topics[0] == ADD_LIQUIDITY_TOPIC:
+            return self._decode_add_liquidity_event(context)
+        elif context.tx_log.topics[0] == REMOVE_LIQUIDITY_TOPIC:
+            return self._decode_remove_liquidity_event(context)
+
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_add_liquidity_event(self, context: DecoderContext) -> DecodingOutput:
+        deposited_amount = asset_normalized_value(
```

Comment:
add a docstring explaining what type of deposit this one is. I guess for the pool?